File: exporter-service/src/routes/exporter/controller/exporterController.ts
import type { Request, Response } from "express";
import type { WordPayloadRequestDTO } from "../dto/request/WordPayloadRequestDTO.js";
import { GenerateWordFileUsecase } from "../usecase/generateWordFile.js";

export async function exportAsWord(req: Request, res: Response) {
  const body: WordPayloadRequestDTO = req.body;

  const usecase = new GenerateWordFileUsecase();

  const buffer: Buffer = await usecase.execute(body);

  const fileName = encodeURIComponent(body.examName || "export") + ".docx";
  res.setHeader(
    "Content-Type",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  );
  res.setHeader(
    "Content-Disposition",
    `attachment; filename*=UTF-8''${fileName}`,
  );

  return res.send(buffer);
}

File: backend/src/main/java/com/wedu/exam_creation/exporter/controller/ExporterController.java
package com.wedu.exam_creation.exporter.controller;

import com.wedu.exam_creation.exporter.dto.request.ExportRequestDTO;
import com.wedu.exam_creation.exporter.usecase.ExporterUsecase;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

@RestController
@RequestMapping("/exporter")
public class ExporterController {
    private final ExporterUsecase exporterUsecase;

    public ExporterController(ExporterUsecase exporterUsecase) {
        this.exporterUsecase = exporterUsecase;
    }

    @PostMapping("/exam")
    public ResponseEntity<Resource> exportWord(@RequestBody ExportRequestDTO payload) {
        byte[] buffer = exporterUsecase.exportAsWord(payload.getExamId());

        ByteArrayResource resource = new ByteArrayResource(buffer);

        String fileName = URLEncoder.encode(payload.getExamName(), StandardCharsets.UTF_8) + ".docx";

        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename*=UTF-8''" + fileName)
                .contentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.wordprocessingml.document"))
                .contentLength(buffer.length)
                .body(resource);
    }
}
